### ISSUE TITLE
Support for setup behind a (HTTPs) load-balancer

### ISSUE THREAD

**Issue by rastapasta** [klokantech/tileserver-gl#85]
Hey klokantech team, thanks for your continuous great work on the server & osm2vectortiles!

Following issue:

When running the tileserver behind a load balancer, the **utils.getTileUrls** (and all other URL rewriting helpers) are generating invalid URLs.

Currently they get generated by combining ``req.protocol + '://' + req.headers.host``
This is totally fine, as long as the tileserver is the endpoint itself.

In my use case, i got

``https://loadbalancer/foo/[*]`` which get's forwarded to one of the HTTP tile servers ``http://192.168.0.x/[*]``

The problem now: **req.protocol** will be **http** (as lb -> tileserver uses it), **req.headers.host** the cluster-internal hostname of the pool's tileserver which got the request. 
 
Long story short: all paths will be generated for the internal host, with internal non-https, instead of pointing it correctly to the loadbalancer's https endpoint.

Would love to propose a new config option named baseURL, which, when set, is being used instead of ``req.protocol + '://' + req.headers.host``.

PR coming & Cheers!

**Comment by petrsloup** [comment]
@rastapasta Isn't this kind of a duplicate to https://github.com/klokantech/tileserver-gl/issues/50 ? Please take a look at https://github.com/klokantech/tileserver-gl/issues/50#issuecomment-249546189

**Comment by rastapasta** [comment]
@petrsloup irks - haven't seen that issue before when looking threw them 😮 
in terms of host and protocol it is a duplicate indeed.

though it still brings one advantage: the server isn't bound to the HTTP root anymore, it can be served from any path via a reverse proxy. not sure if you need that functionality - if so, its there now! if not - all good as well 😄

**Comment by petrsloup** [comment]
@rastapasta I added a short paragraph to the documentation to document the usage of `X-Forwarded-*` headers (http://tileserver.readthedocs.io/en/latest/deployment.html#running-behind-a-proxy-or-a-load-balancer) to help others who might have the same issue.

Serving from non-root paths is not a priority at the moment and the PR would slightly decrease the code readability, so I'm not going to merge it now (although it might come in handy in the future).

**Comment by bergtwvd** [comment]
I know the issue is closed above, but I still like to make an argument to enable others ways for the tileserver to get this information.

I am running the tileserver behind nginx on Kubernetes, using https. Several services are routed via the same address/port, and hence I use a path-prefix to access them. Nginx does not pass on the `X-Forwarded-*` data and I had to request the system admin to change the config of nginx (i.e. adding `enable-x-forwarded` as described at e.g. https://marcolenzo.eu/2019/12/01/enable-x-forwarded-headers-in-kubernetes-nginx-ingress-controller). Having done that it still does not work. Option 2 is to create another ingress on another cluster-node, specifically for the tile server using the root-path to serve data from. Something like a base-url would have solved the problem for me.

**Comment by scara** [comment]
Hi @bergtwvd,
in the mean time, a quick workaround is to "hard-wire" that value via [`-u|--public_url <url>`](https://github.com/maptiler/tileserver-gl/blob/29d3e72dd312e6b367c5adaf46f4f5fad2801174/src/main.js#L48).

HTH,
Matteo

**Comment by bergtwvd** [comment]
Thanks, the `-u` option solved my issue.
